experimental/trusted_shuffler/trusted_shuffler/src/lib.rs

```diff
@@ -16,4 +16,147 @@
+
+    // Async function that sends a request to the backend and returns a response.
+    request_handler: Arc<dyn RequestHandler>,
+}
+
+impl TrustedShuffler {
+    pub fn new(k: usize, request_handler: Arc<dyn RequestHandler>) -> Self {
+        Self {
+            k,
+            requests_to_shuffle: Arc::new(Mutex::new(vec![])),
+            request_handler,
+        }
+    }
+
+    // Asynchronously handles an incoming request.
+    pub async fn invoke(&self, request: Request) -> anyhow::Result<Response> {
+        let (response_sender, response_receiver) = oneshot::channel();
+
+        // Check if the request batch is filled and create a separate task for shuffling requests.
+        {
+            let mut requests_to_shuffle = self
+                .requests_to_shuffle
+                .lock()
+                .map_err(|error| anyhow!("Couldn't lock current messages mutex: {:?}", error))?;
+
+            let request = Message {
+                index: requests_to_shuffle.len(),
+                data: request,
+                response_sender,
+            };
+
+            requests_to_shuffle.push(request);
+
+            if requests_to_shuffle.len() >= self.k {
+                // Replace current requests with an empty vector.
+                let requests = take(requests_to_shuffle.deref_mut());
+
+                // Shuffle requests and spawn async tasks for sending them to the backend.
+                let request_handler = self.request_handler.clone();
+                tokio::spawn(async move {
+                    if let Err(error) =
+                        TrustedShuffler::shuffle_requests(requests, request_handler).await
+                    {
+                        log::error!("Shuffling error: {:?}", error);
+                    }
+                });
+            }
+        }
+
+        response_receiver.await.context("Couldn't receive response")
+    }
+
+    // Lexicographically sorts requests and sends them to the backend using the
+    // [`TrustedShuffler::request_handler`].
+    async fn shuffle_requests(
+        mut requests: Vec<Message>,
+        request_handler: Arc<dyn RequestHandler>,
+    ) -> anyhow::Result<()> {
+        requests.sort_by(|first, second| first.data.cmp(&second.data));
+
+        let mut response_futures: FuturesOrdered<_> = requests
+            .into_iter()
+            .map(|request| {
+                let request_handler_clone = request_handler.clone();
+                async move {
+                    let response = request_handler_clone.handle(request.data).await;
+                    Message {
+                        index: request.index,
+                        data: response,
+                        response_sender: request.response_sender,
+                    }
+                }
+            })
+            .collect();
+        let mut responses = vec![];
+        while let Some(response) = response_futures.next().await {
+            responses.push(response);
+        }
```

Comment:
I think you can use https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.collect here?

Reply:
Done